### DISCUSSION TITLE
Storing Raffle Questions

### DISCUSSION THREAD

**Discussion by erthium** [Duck-Tech-Dev/api-winnerduck#12] [Ideas]
Related: #15

We need to decide how to store raffle questions, how to send that data to front-end and display to the participants.

Suggested structure:

Since we do not how many questions any raffle will have, we need to store question information is a JSON object. We can imagine is as an array of questions, and the question interface could be as below.

**Question**
- Title
- Description
- Answer type (text input, number)
- Preset (email, phone)

Then when the participant has entered the ID of raffle and got in, we can request questions array from the database, which would be in a public route, deserialise the JSON object and display the questions accordingly.

**Comment by HuseyinSimsek7904** [comment]
These are the notes I took while thinking about this:

# Allowed types for form questions:
  ### any (any)                         [shows a textbox that only allows the matched strings]
  - regex restriction
  - minimum length restriction
  - maximum length restriction
    
  ### integer box (integer)             [shows a textbox that only allows integers in the specified range]
  - minimum restriction
  - maximum restriction

  ### decimal box (decimal)             [shows a textbox that only allows decimal numbers in the specified range]
  - minimum restriction
  - maximum restriction
    
  ### slider (slider)                   [shows a textbox that only allows decimal numbers in the specified range]
  - *minimum restriction
  - *maximum restriction
  - steps

  ### yes or no (bool)                  [shows a checkbox in the form]
  ### radio button (radio)              [shows a selection list]
  - list of options
  
  ### phone number (phone_number)       []
  ### email address (email)             []

> - every question in form has the title, description and optional fields
> - "*" indicates required

# A possible JSON questions form:
```
{
  "email": {
    "title": "Email: ",
    "type": "email",
  },
  
  "phone_number": {
    "title": "Phone No: ",
    "optional": true,
    "type": "phone_number",
  },

  "name": {
    "title": "Name: "
    "type": "any",
    "minimum": 3,
    "maximum": 32,
  },
  
  "age": {
    "title": "Age group: "
    "type": "radio",
    "options": [
      "less than 18",
      "between 18-65",
      "more than 65",
    ]
  },
}
```

**Comment by erthium** [comment]
I have some questions but overall looks pretty cool :+1: it's better than what I was imagining in the beginning, i'll have a second look tomorrow, thanks @HuseyinSimsek7904

**Comment by erthium** [comment]
@HuseyinSimsek7904 here is my modified suggestion.
In Typescript we can easily check if a value is given in the question, so if the question has min/max values, we can check it, but if it's not specified, the values taken from JSON will be undefined, so it's cool

```json
[ # changing from object map to an array, we do not need to hold questions as key-value pair
  {
    "title": "Name: "
    "description": "This will be publicly shared in the screen", # this part is added to all
    "optional": false, # optional should be in every question
    "type": "text",  # instead of any
    "minimum": 3,
    "maximum": 32,
  },
  {
    "title": "Email: ",
    "optional": false,
    "description": "Important to contact you in case you are the winner duck!",
    "type": "email",
  },
  {
    "title": "Phone No: ",
    "optional": false,
    "description": "Will not be shared with any third party source",
    "type": "phone",
  },  
  {
    "title": "Age group: "
    "optional": true,
    "description": "Let's us know you better",
    "type": "radio",
    "options": [
      "less than 18",
      "between 18-65",
      "more than 65",
    ]
  },
  { # this question is added to introduce slider
    "title": "How happy are you with the form?"
    "optional": true,
    "description": "Your opinion matters to us",
    "type": "slider",
    "minimum": 1,
    "maximum": 10,
    "step": 1,
  },
]
```

**Reply by HuseyinSimsek7904** [reply]
I was intending to make optional, optional where if it is not added to the question it is meant that the question must be solved. But I don't think it matters, select which one looks the best.

**Reply by erthium** [reply]
Yeah actually, that's better, if the optional is defined and true, then the question will be optional, otherwise will be required as default :+1:

**Comment by erthium** [comment]
If the question storing type is determined, I can arrange the PR in the frontend accordingly and close the issue, what do you guys say @HuseyinSimsek7904 @Utkub24 ?

**Reply by HuseyinSimsek7904** [reply]
I think that makes sense
